### PR TITLE
Camligorkem 3

### PR DESCRIPTION
**Pull request by camligorkem**

### CODE REVIEW

File: camligorkem/assignment3/Dictionary.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
+// Trie Node class to create Dictionary based on prefixes
+class TrieNode{
+    TrieNode[] letters;
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/Dictionary.java]: Maybe better use a map here, this way you won't have to store space for 26 children for every node that you create.

File: camligorkem/assignment3/Dictionary.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        pref.isWord=true;
+    }
+
+    /**
+     * Checks whether a given string is a word in the dictionary.
+     * @param str is the string that we want to know whether a word or not
+     */
+    public boolean isWord(String str) {
+        TrieNode p = searchNode(str);
+        if(p != null) {
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/Dictionary.java]: You can replace all this below with simple return  (p != null && p.isWord);

File: camligorkem/assignment3/DictionaryTest.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import org.junit.Before;
+import org.junit.Test;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Optional;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class DictionaryTest {
+    WordSearch ws;
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/DictionaryTest.java]: Remove unused

File: camligorkem/assignment3/DictionaryTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    Dictionary dic;
+    Optional<HashSet<String>> answer;
+    char[][] grid = {{'a', 'a', 'r'}, {'t', 'c', 'd'}};
+
+    @Before
+    public void setUp() throws Exception {
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/DictionaryTest.java]: This doesn't really throw an exception  :)

File: camligorkem/assignment3/Dictionary.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether a given string is a prefix in the dictionary.
+     * @param prefix is the string that we want to know whether is a prefix or not
+     */
+    public boolean isPrefix(String prefix) {
+        TrieNode p = searchNode(prefix);
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/Dictionary.java]: Just return searchNode(prefix) != null here.

File: camligorkem/assignment3/Dictionary.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            return true;
+        }
+    }
+
+    /**
+     * Search dictionary to see whether a trienode exists for given string
+     * (could be either word or prefix)
+     * @param s is the string to be search in the tree to see if it exists
+     */
+    public TrieNode searchNode(String s){
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/Dictionary.java]: Make this private -- this is only a helper function for isWord and isPrefix.

File: camligorkem/assignment3/Dictionary.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        for(int i=0; i <s.length(); i++){
+            char c = s.charAt(i);
+            int ind = c-'a';
+            if(r.letters[ind]!= null){
+                r = r.letters[ind];
+            }else{
+                return null;
+            }
+        }
+        if(r == root)
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/Dictionary.java]: return r == root ? null : r;  :)

File: camligorkem/assignment3/DictionaryTest.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        assertTrue(dic.isWord("card"));
+        assertTrue(dic.isWord("cat"));
+        assertFalse(dic.isWord("ex"));
+        assertFalse(dic.isWord("false"));
+        assertFalse(dic.isWord(""));
+
+        //isPrefix
+        assertTrue(dic.isPrefix("c"));
+        assertTrue(dic.isPrefix("ca"));
+        assertTrue(dic.isPrefix("c"));
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/DictionaryTest.java]: You already tested that two lines up, remove.

File: camligorkem/assignment3/WordSearch.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+     * @param j columns jth position in the grid to be checked
+     * @param used to see if the given letter in the position is already used (to handle the 3rd condition)
+     * @param prefix the currentPrefix we have to check for the current call of the function.
+     *     Conditions to form a word:
+     *     1. You can start at any position.
+     *     2. You can move to one of the adjacent cells (horizontally/vertically/diagonally).
+     *     3. You can't visit the same cell twice in the same word
+     * @return return a hashset of strings that are found to be words (isWord = true) starting from a certain position
+     */
+    public static Optional<HashSet<String>> findWordsHelper(Dictionary dictionary, char[][] grid, int i, int j, boolean[][] used, String prefix){
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/WordSearch.java]: This should better be private.

File: camligorkem/assignment3/WordSearch.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+public class WordSearch {
+
+    /**
+     * Given a grid and a dictionary it finds the words (isWord=true) that exists in the dictionary
+     * and returns hashset of strings that are words
+     * @param dictionary is made up a list of entries which are either words or prefixes
+     * @param grid the board with the allowed letters
+     * @return return a hashset of strings that are found to be words (isWord = true)
+      */
+    public static Optional<HashSet<String>> findWords(Dictionary dictionary, char[][] grid ){
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/WordSearch.java]: I think Optional here is unnecessary. Returning an empty set should be completely fine. Or do you see any specific benefit in having optional empty?

File: camligorkem/assignment3/WordSearch.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+        int n1 = grid.length;
+        int n2 = grid[0].length;
+        boolean[][] used;
+
+        // HashSet doesn't allow duplicate words, it is a good option to keep the found word
+        Optional<HashSet<String>> foundWords = Optional.empty();
+        Optional<HashSet<String>> res;
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/WordSearch.java]: I don't really see a reason to keep both of those variables. Why won't you just keep adding results of findWordsHelper runs to the foundWords set?

File: camligorkem/assignment3/WordSearch.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+                set.add(currWord);
+                group = Optional.of(set);
+            }
+            //System.out.println("added "+currWord);
+        }
+
+        if(dictionary.isPrefix(currWord)){
+
+            // to make sure we don't exceed the upper or lower bound position of the grid and used board
+            boolean colUpLimit = (j+1) < grid[0].length ? true:false;
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/WordSearch.java]: You don't need "true:false", this is basically colUpLimit = (j+1) < grid[0].length;

File: camligorkem/assignment3/WordSearch.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+            boolean rowUpLimit = (i+1) < grid.length ? true:false;
+            boolean colDownLimit = (j-1) >= 0 ? true:false;
+            boolean rowDownLimit = (i-1) >= 0 ? true:false;
+
+            // --  VERTICAL ---
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/WordSearch.java]: You can define a static array of deltas (or maybe two for two coordinates) and then iterate through them. Then you can change those 8 statements into one loop.

File: camligorkem/assignment3/WordSearch.java (hunk 10 above)
Comment (camligorkem) [reply camligorkem/assignment3/WordSearch.java]: I actually used Optional to have a hands-on practice for me.  It is still a bit blurry for me when I should use optional and when not. If you want I can exclude the optional parts.

File: camligorkem/assignment3/Dictionary.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
+import java.util.HashMap;
+import java.util.Map;
+
+// Trie Node class to create Dictionary based on prefixes
+class TrieNode{
+    Map<Integer, TrieNode> letters;
```
Comment (kkasiacodeu) [inline camligorkem/assignment3/Dictionary.java]: Now you can as well map character not integers, it will make your code much more readable.

**Pull request merged**